src/Mono.Android/Microsoft.Android.Runtime/ManagedValueManager.cs

```diff
@@ -258,44 +243,98 @@ public override List<JniSurfacedPeerInfo> GetSurfacedPeers ()
 			var peers = new List<JniSurfacedPeerInfo> (RegisteredInstances.Count);
 			foreach (var e in RegisteredInstances) {
 				foreach (var p in e.Value) {
-					if (p.Target is not IJavaPeerable peer)
-						continue;
-					peers.Add (new JniSurfacedPeerInfo (e.Key, new WeakReference<IJavaPeerable> (peer)));
+					if (p.Target is IJavaPeerable peer) {
+						peers.Add (new JniSurfacedPeerInfo (e.Key, new WeakReference<IJavaPeerable> (peer)));
+					}
 				}
 			}
 			return peers;
 		}
 	}
 
-	static GCHandle CreateReferenceTrackingHandle (IJavaPeerable value) =>
-		JavaMarshal.CreateReferenceTrackingHandle (value, value.JniObjectReferenceControlBlock);
-		
-	static unsafe void FreeHandle (GCHandle handle)
+	unsafe struct WeakPeerReference : IDisposable
+	{
+		private WeakReference<IJavaPeerable> _weakReference;
+		private GCHandle _handle;
+
+		public WeakPeerReference (IJavaPeerable peer)
+		{
+			_weakReference = new WeakReference<IJavaPeerable> (peer);
+
+			var handleContext = (HandleContext*) NativeMemory.AllocZeroed (1, HandleContext.Size);
+			if (handleContext == null) {
+				throw new OutOfMemoryException ("Failed to allocate memory for HandleContext.");
+			}
+
+			_handle = JavaMarshal.CreateReferenceTrackingHandle (peer, handleContext);
+
+			handleContext->Handle = GCHandle.ToIntPtr (_handle);
+			handleContext->ControlBlock = peer.JniObjectReferenceControlBlock;
+		}
+
+		public void FreeContext ()
+		{
+			var context = (HandleContext*) JavaMarshal.GetContext (_handle);
+			if (context != null) {
+				NativeMemory.Free (context);
+			}
+		}
+
+		public void Dispose ()
```

Comment:
Note that this should be legal to call only if the caller holds a reference to the `Target`.